### ISSUE TITLE
Keep docker image cache in sync with station profile

### ISSUE THREAD

**Issue by porst17** [hilbert/hilbert-cli#2]
When a station is provisioned (currently called `prepared`, but I suggest to change the term), it downloads all docker images it needs during regular operation. The available docker images must be in sync with the station profile, i.e. whenever the profile is updated, the station might be _reprovisioned_. New images have to be downloaded, outdated images have to be deleted.

In my opinion, this is the easiest solution. If you never delete images they might pile up (e.g. due to many versions of the same exhibit) and the station might run out of space and become out of order.

**Comment by elondaits** [comment]
It would be nice if it didn't delete any images it might need when it's provisioned. That would save from re-downloading the OS, at minimum, right?

As a future enhancement a cache-like logic could be used:
- A quota (as a percentage of the HDD capacity or an absolute value) is defined.
- When provisioning if downloading images would go over the quota, make space either by:
  - Deleting everything
  - Delete everything except images needed for the current provisioning
  - Delete the images with the oldest dates of last usage that are not needed for the current provisioning until enough space is freed so that `stored images + new images < quota`.

Any of these alternatives (even the first one) would allow for fast switching between 2 or more apps that are frequently used as long as the total size is under the quota.

**Comment by porst17** [comment]
All images should be kept during reprovisioning. It is likely that most of the image layers will be reused which would safe a lot of network bandwidth and time. Once the reprovisioning is done, you could delete all images that are not needed anymore. This way, you would store at most two sets of images: the ones needed before the update and the ones needed afterwards. We should have sufficient space for this solutions. If downloading the second set of images fails due to space limitations, you could also delete all images and start from scratch.

BTW: AFAIK it is not so easy to know the space an image consumes when unpacked on disk, i.e. it might not be trivial to fill in the numbers for @elondaits's cache cleanup formula.

**Comment by malex984** [comment]
Recall that we currently use docker-compose to pull docker images required to run some services.
It will be quite complicated to estimate necessary disk space for a service as it depends on which docker image is used for it as well as what was previously pulled to the station. It can only be done once it is pulled to the station OR (very roughly) on the server side (see http://tuhrig.de/flatten-a-docker-container-or-image/ , http://stackoverflow.com/questions/26753087/docker-how-to-analyze-a-containers-disk-usage).

Moreover there is no  usage date attached to docker images... but that may be emulated additionally by our scripts.

Currently i am planning to pull everything and remove outdated (untagged) images with `docker rmi $(docker images | grep "^<none>" | awk '{print $3}')` according to http://jimhoskins.com/2013/07/27/remove-untagged-docker-images.html

Additionally i think it may be possible to figure out all the necessary docker images (according to the current profile) and remove the rest of local images - assuming there can be no separate data-only images...

**Comment by malex984** [comment]
For the reference: https://github.com/spotify/docker-gc
